superset/common/query_actions.py

```diff
@@ -0,0 +1,182 @@
+) -> Dict[str, Any]:
+    datasource = _get_datasource(query_context, query_obj)
+    return {
+        "query": datasource.get_query_str(query_obj.to_dict()),
+        "language": datasource.query_language,
+    }
+
+
+def _get_full(
+    query_context: "QueryContext",
+    query_obj: "QueryObject",
+    force_cached: Optional[bool] = False,
+) -> Dict[str, Any]:
+    datasource = _get_datasource(query_context, query_obj)
+    result_type = query_obj.result_type or query_context.result_type
+    payload = query_context.get_df_payload(query_obj, force_cached=force_cached)
+    df = payload["df"]
+    status = payload["status"]
+    if status != QueryStatus.FAILED:
+        payload["colnames"] = list(df.columns)
+        payload["coltypes"] = extract_dataframe_dtypes(df)
+        payload["data"] = query_context.get_data(df)
+    del payload["df"]
+
+    filters = query_obj.filter
+    filter_columns = cast(List[str], [flt.get("col") for flt in filters])
+    columns = set(datasource.column_names)
+    applied_time_columns, rejected_time_columns = get_time_filter_status(
+        datasource, query_obj.applied_time_extras
+    )
+    payload["applied_filters"] = [
+        {"column": col} for col in filter_columns if col in columns
+    ] + applied_time_columns
+    payload["rejected_filters"] = [
+        {"reason": "not_in_datasource", "column": col}
+        for col in filter_columns
+        if col not in columns
+    ] + rejected_time_columns
+
+    if result_type == ChartDataResultType.RESULTS and status != QueryStatus.FAILED:
+        return {"data": payload["data"]}
+    return payload
+
+
+def _get_samples(
+    query_context: "QueryContext", query_obj: "QueryObject", force_cached: bool = False
+) -> Dict[str, Any]:
+    datasource = _get_datasource(query_context, query_obj)
+    row_limit = query_obj.row_limit or math.inf
+    query_obj = copy.copy(query_obj)
+    query_obj.is_timeseries = False
+    query_obj.orderby = []
+    query_obj.groupby = []
+    query_obj.metrics = []
+    query_obj.post_processing = []
+    query_obj.row_limit = min(row_limit, config["SAMPLES_ROW_LIMIT"])
+    query_obj.row_offset = 0
+    query_obj.columns = [o.column_name for o in datasource.columns]
+    return _get_full(query_context, query_obj, force_cached)
+
+
+def _get_results(
+    query_context: "QueryContext", query_obj: "QueryObject", force_cached: bool = False
+) -> Dict[str, Any]:
+    payload = _get_full(query_context, query_obj, force_cached)
+    return {"data": payload["data"]}
+
+
+_result_type_functions: Dict[
+    ChartDataResultType, Callable[["QueryContext", "QueryObject", bool], Dict[str, Any]]
+] = {
+    ChartDataResultType.COLUMNS: _get_columns,
+    ChartDataResultType.TIMEGRAINS: _get_timegrains,
+    ChartDataResultType.QUERY: _get_query,
+    ChartDataResultType.SAMPLES: _get_samples,
+    ChartDataResultType.FULL: _get_full,
+    ChartDataResultType.RESULTS: _get_results,
```

Comment:
@villebro Sorry for being late to the party, but it there a reason why `is_rowcount` is not implemented as another `ChartDataResultType`?

Reply:
That's the way I thought about implementing it first, but I felt it better to implement it as a `QueryObject` property, as it has to be passed to the SQLA model as it affects the rendered query (I feel `QueryObject` should be mostly 1-1 mapped to the query dict that's unpacked into `get_sqla_query`).

Reply:
Wouldn't you still have that mapping if you just create a `_get_rowcount` function that sets `QueryObjects`'s `is_rowcount` to `True`?  I see no difference in setting `is_rowcount` in a `_get_rowcount` function than manipulating other fields of `QueryObject` in `_get_samples` and `_get_query`.

The reason why I think a new result type makes more sense is that from the client-side point of view,  `ResultType.SAMPLES`, `ResultType.QUERY`, and `ResultType.ROW_COUNT` are mutually exclusive, i.e., you cannot have a `QueryObject` with both `result_type = ResultType.SAMPLES` and `is_rowcount = True`.

Reply:
I don't mind adding a result type as a convenience method for fetching the row count. However, the preferred method for retrieving row counts should IMO be explicitly setting `is_rowcount` on `QueryObject`, as it will make it possible to set `ResultType.QUERY` on `QueryContext` and by doing so fetch only the queries for all `QueryObject`s (we don't yet support showing queries for multiple queries, but I hope we can add it soon): https://github.com/apache/superset/blob/ca27b00f3a2ce6f1a2b4f2e4a6586b3c3fc578a5/superset-frontend/src/explore/components/DisplayQueryButton.jsx#L170